### PR TITLE
Deprecate `DetermineError` functions

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area ops-productivity quality
/kind technical-debt enhancement
/merge squash

**What this PR does / why we need it**:
This PR is the first step towards solving the issue #5444. 
Currently, we have calls to `gardencorev1beta1helper.DetermineError` from the extension library which uses a centrally maintained regex to determine the Gardener Error Codes. We want to move away from this mechanism and have the extensions figure out the error codes themselves. Extensions should have their own detection/mapping logic.
~~This PR adds a new interface ErrorCodeDetector to the extension controllers.~~
~~Next step is to add the error handling in extensions after vendoring a gardener version with this PR, which would look similar to this: https://github.com/gardener/gardener-extension-provider-aws/compare/master...shafeeqes:enh/ReworkErrorCodes-3~~
~~And the final step is to make changes in g/g to adapt this error handling to determine the error codes, which is added as TODO.~~

This approach was dropped since it introduces more complexity and makes the code hard to follow.

So, Presently, this PR deprecates the DetermineError and DetermineErrorCodes functions to be removed in a future version.
Extensions should have their own error code mappings and should return the corresponding error codes to Gardener. ie, gardenlet should only pick up the status.lastError.codes list populated by the extension objects. 

**Which issue(s) this PR fixes**:
Fixes part of #5444

**Special notes for your reviewer**:

**Release note**:

```other operator
The generic error code mapping in Gardener is deprecated. Extensions should use their own error code mappings and should return corresponding error codes to Gardener.
```

### CODE REVIEW

File: extensions/pkg/controller/backupbucket/errorcodedetector.go
Comment (ialidzhikov) [inline extensions/pkg/controller/backupbucket/errorcodedetector.go]: What was the motivation to duplicate this interface n times for each extension kind pkg? I would rather vote to have it centrally in one place under the extension library.

Comment (ialidzhikov) [inline extensions/pkg/controller/backupbucket/errorcodedetector.go]: You can also enhance the PR description and provider details about the motivation for this change (that currently we have calls to `gardencorev1beta1helper.DetermineError` from the extension library and we need a mechanism for extensions to provider their own error code mapping/detection).
You could also link an example change on how the adaption looks like on the extension side.

Comment (shafeeqes) [reply extensions/pkg/controller/backupbucket/errorcodedetector.go]: How about `extensions/pkg/controller/common` or `extensions/pkg/util`?

Comment (ialidzhikov) [reply extensions/pkg/controller/backupbucket/errorcodedetector.go]: IMO `extensions/pkg/controller/common/error_code_detector.go` would be fine.

Comment (ialidzhikov) [reply extensions/pkg/controller/backupbucket/errorcodedetector.go]: Hm, I now see that we have 1 usage of `DetermineError` in `extensions/pkg/terraformer` that is not adapted with this PR. Then IMO it would be better to place it under `extensions/pkg/util/error_code_detector.go` as it won't be used from `extensions/pkg/controller` only.

Comment (shafeeqes) [reply extensions/pkg/controller/backupbucket/errorcodedetector.go]: I will add an example with provider-aws soon.

Comment (shafeeqes) [reply extensions/pkg/controller/backupbucket/errorcodedetector.go]: It would look like this: https://github.com/gardener/gardener-extension-provider-aws/compare/master...shafeeqes:enh/ReworkErrorCodes-3

**Comment by shafeeqes** [comment]
Can you PTAL? @ialidzhikov

File: extensions/pkg/controller/backupbucket/controller.go
Comment (timebertt) [inline extensions/pkg/controller/backupbucket/controller.go]: TBH, I don't get the motivation behind introducing this `ErrorCodeDetector` interface in each and every controller in the extensions library. This new touches a lot of places that hard to follow already as of now.

If your goal is to have a central error code detection mechanism per extension, I would rather go for wrapping `reconcile.Reconciler` and just ran the returned error (if any) through the detection.
In the extensions library, there should be a check whether the returned error contains an error code (for example, by using `errors.As` with `gardencorev1beta1helper.Coder` interface). If such code was attached to the error, then the controller in the library should set the status accordingly.

**Review by timebertt — COMMENTED** [review]
Sorry for the late feedback, this slipped through on my end.

I'm not sure, I can follow your approach. Maybe we can have a call and discuss some ideas.
/assign

File: pkg/apis/core/v1beta1/helper/errors.go
Comment (timebertt) [inline pkg/apis/core/v1beta1/helper/errors.go]: This should be `Unwrap`, see https://go.dev/blog/go1.13-errors

File: pkg/apis/core/v1beta1/helper/errors.go
Comment (timebertt) [inline pkg/apis/core/v1beta1/helper/errors.go]: Just return `e.err.Error()` here.

File: pkg/apis/core/v1beta1/helper/errors.go
Comment (timebertt) [inline pkg/apis/core/v1beta1/helper/errors.go]: This is already covered by `DetermineErrorCodes`. I think, this can simply be dropped.

File: pkg/apis/core/v1beta1/helper/errors.go
Comment (timebertt) [inline pkg/apis/core/v1beta1/helper/errors.go]: `len(codes) == 0` is the idiomatic style

File: extensions/pkg/controller/status.go
Comment (timebertt) [inline extensions/pkg/controller/status.go]: I think, we should already drop `DetermineError` usage here, as the shoot flow tasks that wait for extension resources will still detect error codes from the error message.
We should only keep it there for a while and then drop it there (as outlined in https://github.com/gardener/gardener/issues/5444).

**Review by timebertt — COMMENTED** [review]
This looks better now, thanks for adapting.
I have a few more suggestions in line to simplify things :)

File: pkg/apis/core/v1beta1/helper/errors.go
Comment (shafeeqes) [reply pkg/apis/core/v1beta1/helper/errors.go]: Yes, the detection of codes is fine. 
The DetermineError is already called above this.
But the ErrorWithCodes now will have duplicated fields.
Something like,
```
Expected
      <*helper.ErrorWithCodes | 0xc0005495c0>: {
          err: <*fmt.wrapError | 0xc00038efa0>{
              msg: "unauthenticated",
              err: <*helper.ErrorWithCodes | 0xc000382a20>{
                  err: <*errors.errorString | 0xc0003879f0>{s: "unauthenticated"},
                  codes: [
                      "ERR_INFRA_UNAUTHENTICATED",
                  ],
              },
          },
          codes: [
              "ERR_INFRA_UNAUTHENTICATED",
          ],
      }
  to equal
      <*helper.ErrorWithCodes | 0xc000382a50>: {
          err: <*errors.errorString | 0xc000387a10>{s: "unauthenticated"},
          codes: [
              "ERR_INFRA_UNAUTHENTICATED",
          ],
      }
```

File: extensions/pkg/controller/status.go
Comment (shafeeqes) [reply extensions/pkg/controller/status.go]: > as the shoot flow tasks that wait for extension resources will still detect error codes from the error message.

hmm.. where is this done exactly? 
I couldn't find a usage of DetermineError or DetermineErrorCodes corresponding to it.

**Review by timebertt — APPROVED** [review]
Thank you very much, this looks good now.
/lgtm

**Comment by ialidzhikov** [comment]
Currently this PR (not in the v1.46 milestone) blocks prow and prevents merging PRs (that are in the v1.46 milestone). Looks like prow thinks this PR is ready to merge but then it fails due to the release milestone check.
I will cancel the approval for now to unblock other PRs. @timebertt you can add the approval again once the v1.46 release is cut.

~~/approve cancel~~

EDIT: looks like `/approve cancel` works only for your own approvals. Anyways, Oliver already added a `/hold` label.

**Comment by oliver-goetz** [comment]
/hold
this PR is blocking tide because "Release Milestand Check" is failing

**Comment by timebertt** [comment]
I think, this was a race between the release milestone check (triggered by prow adding `approved`/`lgtm`) and tide's sync.

**Comment by oliver-goetz** [comment]
/unhold
Test if there have been a race

**Comment by oliver-goetz** [comment]
/hold
commit [a0042f8](https://github.com/gardener/gardener/pull/5625/commits/a0042f81bc864a847a5822dcb6ed9a5d3c046ce4) has the wrong status (success) of `Release Milestone Check`. It seems that it was not updated.
That's why Tide is trying to merge this PR when it is not _hold_

**Comment by ialidzhikov** [comment]
/hold cancel
as v1.46.0 was released earlier today

**Pull request merged**
